OpenSim/Sandbox/ExampleHopperDevice/exampleHopperDevice.cpp

```diff
@@ -1,428 +1,285 @@
-    }
-private:
-    void constructProperties() {
-        constructProperty_function(Constant(0.0));
-    }
-}; // end of SignalGenerator
-
-
-
-/*************** HELPER FUNCTIONS: Implemented in answers.cpp ****************/
- // Utility to load and draw a model in a refresh loop
- // so that edits to the modelFile can be visualized immediately.
-void refreshModel(const std::string& modelFile);
-
-// Helper method to edit a device's path (if it has one) to handle
-// wrapping over a wrap surface already in the model.
-void handlePathWrapping(OpenSim::ModelComponent* device,
-    OpenSim::Model& model);
-
-// Main driver to simulate a model from an initial state
-// Simulate means to integrate the model equations forward in time.
-// The State is updated so that it returns the final state of integration.
-void simulate(OpenSim::Model& model, SimTK::State& state);
-/************ end of HELPER FUNCTIONS: Implemented in answers.cpp ************/
-
-
-OpenSim::Model createTestBed() {
-    using SimTK::Vec3;
-    using SimTK::Inertia;
+namespace OpenSim {
 
-    OpenSim::Model testBed;
-    testBed.setName("testbed");
-    testBed.setUseVisualizer(true);
-    testBed.setGravity(Vec3(0));
+// Forward declarations for methods used below.
+Model buildHopper();    //defined in buildHopperModel.cpp
+Model buildTestbed();   //defined in helperMethods.h
+Device* buildDevice();  //defined in buildDevice.cpp
 
-    // Get the ground frame which must be the base of any mechanism
-    auto& ground = testBed.getGround();
 
-    // Create a load of mass LOAD kg.
-    auto load = new OpenSim::Body("load", LOAD, Vec3(0), Inertia(1));
-    // Set properties of a sphere geometry to be used for the load.
-    OpenSim::Sphere sphere;
-    sphere.setFrameName("load");
-    sphere.set_radius(0.02);
-    sphere.setOpacity(0.5);
-    sphere.setColor(Vec3{ 0, 0, 1 });
-    load->addGeometry(sphere);
-    testBed.addBody(load);
+//------------------------------------------------------------------------------
+// Attach the device to any two PhysicalFrames in a model.
+// [Step 2, Task D]
+//------------------------------------------------------------------------------
+void connectDeviceToModel(OpenSim::Device& device, OpenSim::Model& model,
+    const std::string& modelFrameAname, const std::string& modelFrameBname)
+{
+    //TODO: Get references to the "anchor" joints in the device.
+
+    //TODO: Recall that the child frame of each anchor (WeldJoint) was attached
+    //      to the corresponding cuff. We will now attach the parent frames of
+    //      the anchors to modelFrameA and modelFrameB. First, get a reference
+    //      to the specified PhysicalFrames in model, then connect them to the
+    //      parent frames of each anchor (2 lines of code for each cuff).
+
+    // Add the device to the model. We need to add the device using
+    // addModelComponent() rather than addComponent() because of a bug in
+    // Model::initSystem().
```

Comment:
BTW I tried merging with master and `addModelComponent()` is still necessary. The issue is not ground--it's that WeldJoint `anchorA` needs Body `load`'s mobilized body index. So it's the more general problem.